{
 "metadata": {
  "name": "",
  "signature": "sha256:9e9897880cc2b6801d05ad6c212d1ed8b6daf5f9d4e1ef2cac6b311d0c7dfdd4"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from __future__ import unicode_literals\n",
      "import json\n",
      "import numpy as np\n",
      "import pandas as pd\n",
      "from pandas import DataFrame, Series"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Transforming batch-collected article data\n",
      "\n",
      "The desired data structure for article information is the following JSON object:\n",
      "\n",
      "```\n",
      "<doi1>: {\n",
      "    author: [ ... ]\n",
      "    title:\n",
      "    journal:\n",
      "    publication_date: <yyyy>\n",
      "    subject: [ <full subject /-separated strings>, ... ]\n",
      "    subj_top: [ set of top levels of each subject ]\n",
      "    subj_leaf: [ set of last terms of each subject ]\n",
      "},\n",
      "<doi2>: { ... }, \n",
      "...\n",
      "```"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df = pd.read_pickle('../data/all_plos_df.pkl')\n",
      "df.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>abstract</th>\n",
        "      <th>author</th>\n",
        "      <th>id</th>\n",
        "      <th>journal</th>\n",
        "      <th>publication_date</th>\n",
        "      <th>score</th>\n",
        "      <th>subject</th>\n",
        "      <th>title_display</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>0</th>\n",
        "      <td> [\\nClenbuterol, a \u03b22-agonist, induces skeletal...</td>\n",
        "      <td> [Pascal Sirvent, Aymerick Douillard, Olivier G...</td>\n",
        "      <td> 10.1371/journal.pone.0100281</td>\n",
        "      <td>                         PLoS ONE</td>\n",
        "      <td> 2014-06-27T00:00:00Z</td>\n",
        "      <td> 1</td>\n",
        "      <td> [/Medicine and health sciences/Pathology and l...</td>\n",
        "      <td> Effects of Chronic Administration of Clenbuter...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>1</th>\n",
        "      <td> [\\nRecent studies point to an association betw...</td>\n",
        "      <td> [Lan Zhang, Hui Ding, Dan-Hui Wang, Yan-Li Zha...</td>\n",
        "      <td> 10.1371/journal.pone.0070935</td>\n",
        "      <td>                         PLoS ONE</td>\n",
        "      <td> 2013-08-09T00:00:00Z</td>\n",
        "      <td> 1</td>\n",
        "      <td> [/Biology and life sciences/Molecular biology/...</td>\n",
        "      <td> Calpastatin Gene (CAST) Is Not Associated with...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>2</th>\n",
        "      <td> [\\nThe reduction of game and fish populations ...</td>\n",
        "      <td> [Silvia D\u00edaz-Fern\u00e1ndez, Beatriz Arroyo, Fabi\u00e1n...</td>\n",
        "      <td> 10.1371/journal.pone.0066671</td>\n",
        "      <td>                         PLoS ONE</td>\n",
        "      <td> 2013-06-19T00:00:00Z</td>\n",
        "      <td> 1</td>\n",
        "      <td> [/Biology and life sciences/Population biology...</td>\n",
        "      <td> Effect of Game Management on Wild Red-Legged P...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>3</th>\n",
        "      <td> [Background: Cebu has been one of the most lep...</td>\n",
        "      <td> [Pauline F D Scheelbeek, Marivic V F Balagon, ...</td>\n",
        "      <td> 10.1371/journal.pntd.0002444</td>\n",
        "      <td> PLoS Neglected Tropical Diseases</td>\n",
        "      <td> 2013-09-19T00:00:00Z</td>\n",
        "      <td> 1</td>\n",
        "      <td> [/Earth sciences/Geography/Geographic areas/Ru...</td>\n",
        "      <td> A Retrospective Study of the Epidemiology of L...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4</th>\n",
        "      <td> [Background: Understanding how androgen recept...</td>\n",
        "      <td> [Adam T Szafran, Maria Szwarc, Marco Marcelli,...</td>\n",
        "      <td> 10.1371/journal.pone.0003605</td>\n",
        "      <td>                         PLoS ONE</td>\n",
        "      <td> 2008-11-03T00:00:00Z</td>\n",
        "      <td> 1</td>\n",
        "      <td> [/Biology and life sciences/Cell biology/Cell ...</td>\n",
        "      <td> Androgen Receptor Functional Analyses by High ...</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>5 rows \u00d7 8 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 2,
       "text": [
        "                                            abstract  \\\n",
        "0  [\\nClenbuterol, a \u03b22-agonist, induces skeletal...   \n",
        "1  [\\nRecent studies point to an association betw...   \n",
        "2  [\\nThe reduction of game and fish populations ...   \n",
        "3  [Background: Cebu has been one of the most lep...   \n",
        "4  [Background: Understanding how androgen recept...   \n",
        "\n",
        "                                              author  \\\n",
        "0  [Pascal Sirvent, Aymerick Douillard, Olivier G...   \n",
        "1  [Lan Zhang, Hui Ding, Dan-Hui Wang, Yan-Li Zha...   \n",
        "2  [Silvia D\u00edaz-Fern\u00e1ndez, Beatriz Arroyo, Fabi\u00e1n...   \n",
        "3  [Pauline F D Scheelbeek, Marivic V F Balagon, ...   \n",
        "4  [Adam T Szafran, Maria Szwarc, Marco Marcelli,...   \n",
        "\n",
        "                             id                           journal  \\\n",
        "0  10.1371/journal.pone.0100281                          PLoS ONE   \n",
        "1  10.1371/journal.pone.0070935                          PLoS ONE   \n",
        "2  10.1371/journal.pone.0066671                          PLoS ONE   \n",
        "3  10.1371/journal.pntd.0002444  PLoS Neglected Tropical Diseases   \n",
        "4  10.1371/journal.pone.0003605                          PLoS ONE   \n",
        "\n",
        "       publication_date  score  \\\n",
        "0  2014-06-27T00:00:00Z      1   \n",
        "1  2013-08-09T00:00:00Z      1   \n",
        "2  2013-06-19T00:00:00Z      1   \n",
        "3  2013-09-19T00:00:00Z      1   \n",
        "4  2008-11-03T00:00:00Z      1   \n",
        "\n",
        "                                             subject  \\\n",
        "0  [/Medicine and health sciences/Pathology and l...   \n",
        "1  [/Biology and life sciences/Molecular biology/...   \n",
        "2  [/Biology and life sciences/Population biology...   \n",
        "3  [/Earth sciences/Geography/Geographic areas/Ru...   \n",
        "4  [/Biology and life sciences/Cell biology/Cell ...   \n",
        "\n",
        "                                       title_display  \n",
        "0  Effects of Chronic Administration of Clenbuter...  \n",
        "1  Calpastatin Gene (CAST) Is Not Associated with...  \n",
        "2  Effect of Game Management on Wild Red-Legged P...  \n",
        "3  A Retrospective Study of the Epidemiology of L...  \n",
        "4  Androgen Receptor Functional Analyses by High ...  \n",
        "\n",
        "[5 rows x 8 columns]"
       ]
      }
     ],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Drop unused data.\n",
      "df.drop(['author', 'title_display', 'journal', 'abstract', 'score'], axis=1, inplace=True)\n",
      "df.set_index('id', inplace=True)\n",
      "\n",
      "# We just want the year.\n",
      "df.publication_date = df.publication_date.str[:4]\n",
      "\n",
      "df.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>publication_date</th>\n",
        "      <th>subject</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>id</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>10.1371/journal.pone.0100281</th>\n",
        "      <td> 2014</td>\n",
        "      <td> [/Medicine and health sciences/Pathology and l...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>10.1371/journal.pone.0070935</th>\n",
        "      <td> 2013</td>\n",
        "      <td> [/Biology and life sciences/Molecular biology/...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>10.1371/journal.pone.0066671</th>\n",
        "      <td> 2013</td>\n",
        "      <td> [/Biology and life sciences/Population biology...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>10.1371/journal.pntd.0002444</th>\n",
        "      <td> 2013</td>\n",
        "      <td> [/Earth sciences/Geography/Geographic areas/Ru...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>10.1371/journal.pone.0003605</th>\n",
        "      <td> 2008</td>\n",
        "      <td> [/Biology and life sciences/Cell biology/Cell ...</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>5 rows \u00d7 2 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 3,
       "text": [
        "                             publication_date  \\\n",
        "id                                              \n",
        "10.1371/journal.pone.0100281             2014   \n",
        "10.1371/journal.pone.0070935             2013   \n",
        "10.1371/journal.pone.0066671             2013   \n",
        "10.1371/journal.pntd.0002444             2013   \n",
        "10.1371/journal.pone.0003605             2008   \n",
        "\n",
        "                                                                        subject  \n",
        "id                                                                               \n",
        "10.1371/journal.pone.0100281  [/Medicine and health sciences/Pathology and l...  \n",
        "10.1371/journal.pone.0070935  [/Biology and life sciences/Molecular biology/...  \n",
        "10.1371/journal.pone.0066671  [/Biology and life sciences/Population biology...  \n",
        "10.1371/journal.pntd.0002444  [/Earth sciences/Geography/Geographic areas/Ru...  \n",
        "10.1371/journal.pone.0003605  [/Biology and life sciences/Cell biology/Cell ...  \n",
        "\n",
        "[5 rows x 2 columns]"
       ]
      }
     ],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def get_subj_top(subjects):\n",
      "    subj_top = set()\n",
      "    for s in subjects:\n",
      "        # the string gets split at its first character, so not [0] here:\n",
      "        subj_top.add(s.split('/')[1])\n",
      "    return subj_top\n",
      "\n",
      "def get_subj_leaf(subjects):\n",
      "    subj_top = set()\n",
      "    for s in subjects:\n",
      "        subj_top.add(s.split('/')[-1])\n",
      "    return subj_top"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df['subj_top'] = df.subject.apply(get_subj_top)\n",
      "df['subj_leaf'] = df.subject.apply(get_subj_leaf)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 5
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "test_df = df.sort_index().head()\n",
      "test_df"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>publication_date</th>\n",
        "      <th>subject</th>\n",
        "      <th>subj_top</th>\n",
        "      <th>subj_leaf</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>id</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>10.1371/journal.pbio.0000001</th>\n",
        "      <td> 2003</td>\n",
        "      <td> [/Biology and life sciences/Molecular biology/...</td>\n",
        "      <td> set([Biology and life sciences, Physical scien...</td>\n",
        "      <td> set([Recombinant proteins, Nucleotide sequenci...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>10.1371/journal.pbio.0000002</th>\n",
        "      <td> 2003</td>\n",
        "      <td> [/Biology and life sciences/Molecular biology/...</td>\n",
        "      <td> set([Biology and life sciences, Physical scien...</td>\n",
        "      <td> set([DNA sequences, Microarrays, Oligonucleoti...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>10.1371/journal.pbio.0000003</th>\n",
        "      <td> 2003</td>\n",
        "      <td> [/Biology and life sciences/Organisms/Viruses/...</td>\n",
        "      <td> set([Biology and life sciences, Medicine and h...</td>\n",
        "      <td> set([DNA sequences, West Nile virus, Sequence ...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>10.1371/journal.pbio.0000004</th>\n",
        "      <td> 2003</td>\n",
        "      <td> [/Medicine and health sciences/Immunology/Immu...</td>\n",
        "      <td> set([Biology and life sciences, Medicine and h...</td>\n",
        "      <td> set([Recombinant proteins, Antibodies, Immune ...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>10.1371/journal.pbio.0000005</th>\n",
        "      <td> 2003</td>\n",
        "      <td> [/Biology and life sciences/Biochemistry/Prote...</td>\n",
        "      <td> set([Biology and life sciences, Physical scien...</td>\n",
        "      <td> set([Plasmodium, Merozoites, Oligonucleotides,...</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>5 rows \u00d7 4 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 6,
       "text": [
        "                             publication_date  \\\n",
        "id                                              \n",
        "10.1371/journal.pbio.0000001             2003   \n",
        "10.1371/journal.pbio.0000002             2003   \n",
        "10.1371/journal.pbio.0000003             2003   \n",
        "10.1371/journal.pbio.0000004             2003   \n",
        "10.1371/journal.pbio.0000005             2003   \n",
        "\n",
        "                                                                        subject  \\\n",
        "id                                                                                \n",
        "10.1371/journal.pbio.0000001  [/Biology and life sciences/Molecular biology/...   \n",
        "10.1371/journal.pbio.0000002  [/Biology and life sciences/Molecular biology/...   \n",
        "10.1371/journal.pbio.0000003  [/Biology and life sciences/Organisms/Viruses/...   \n",
        "10.1371/journal.pbio.0000004  [/Medicine and health sciences/Immunology/Immu...   \n",
        "10.1371/journal.pbio.0000005  [/Biology and life sciences/Biochemistry/Prote...   \n",
        "\n",
        "                                                                       subj_top  \\\n",
        "id                                                                                \n",
        "10.1371/journal.pbio.0000001  set([Biology and life sciences, Physical scien...   \n",
        "10.1371/journal.pbio.0000002  set([Biology and life sciences, Physical scien...   \n",
        "10.1371/journal.pbio.0000003  set([Biology and life sciences, Medicine and h...   \n",
        "10.1371/journal.pbio.0000004  set([Biology and life sciences, Medicine and h...   \n",
        "10.1371/journal.pbio.0000005  set([Biology and life sciences, Physical scien...   \n",
        "\n",
        "                                                                      subj_leaf  \n",
        "id                                                                               \n",
        "10.1371/journal.pbio.0000001  set([Recombinant proteins, Nucleotide sequenci...  \n",
        "10.1371/journal.pbio.0000002  set([DNA sequences, Microarrays, Oligonucleoti...  \n",
        "10.1371/journal.pbio.0000003  set([DNA sequences, West Nile virus, Sequence ...  \n",
        "10.1371/journal.pbio.0000004  set([Recombinant proteins, Antibodies, Immune ...  \n",
        "10.1371/journal.pbio.0000005  set([Plasmodium, Merozoites, Oligonucleotides,...  \n",
        "\n",
        "[5 rows x 4 columns]"
       ]
      }
     ],
     "prompt_number": 6
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Start making dictionaries\n",
      "\n",
      "We need to reduce the amount of data that the web app needs to load. We will encode all the subject terms and paths into dicts with simple numeric keys."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def make_dicts(values):\n",
      "    # Dict to translate original terms into codes\n",
      "    d1 = {}\n",
      "    index = 1\n",
      "    \n",
      "    for i in values:\n",
      "        for s in i:\n",
      "            s = s.replace(u'\\u2019', \"'\")\n",
      "            if s not in d1.keys():\n",
      "                d1[s] = index\n",
      "                index = index + 1\n",
      "    \n",
      "    # Dict to translate codes back into original terms\n",
      "    d2 = {d1[k]: k for k in d1.keys()}\n",
      "    \n",
      "    return d1, d2\n",
      "\n",
      "\n",
      "def encode_terms(values, d):\n",
      "    # Returned Series must share the same Index as input\n",
      "    coded_values = values.copy()\n",
      "    \n",
      "    for i in range(len(values)):\n",
      "        coded_values.ix[i] = set(d[s.replace(u'\\u2019', \"'\")] for s in values.ix[i])\n",
      "    \n",
      "    return coded_values"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 7
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Test!\n",
      "# - There should be repeated uses of the keys (not just 1, 2, 3, 4, 5...)\n",
      "# - There should be the same number of terms in original and 'encoded' data\n",
      "\n",
      "print test_df['subj_leaf'].apply(len)\n",
      "\n",
      "d_to_code, d_from_code = make_dicts(test_df['subj_leaf'])\n",
      "test_enc1 = encode_terms(test_df['subj_leaf'], d_to_code)\n",
      "print test_enc1\n",
      "print test_enc1.apply(len) == test_df['subj_leaf'].apply(len)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "id\n",
        "10.1371/journal.pbio.0000001    8\n",
        "10.1371/journal.pbio.0000002    8\n",
        "10.1371/journal.pbio.0000003    8\n",
        "10.1371/journal.pbio.0000004    8\n",
        "10.1371/journal.pbio.0000005    8\n",
        "Name: subj_leaf, dtype: int64\n",
        "id\n",
        "10.1371/journal.pbio.0000001            set([1, 2, 3, 4, 5, 6, 7, 8])\n",
        "10.1371/journal.pbio.0000002       set([3, 6, 9, 10, 11, 12, 13, 14])\n",
        "10.1371/journal.pbio.0000003      set([3, 9, 12, 14, 15, 16, 17, 18])\n",
        "10.1371/journal.pbio.0000004      set([1, 3, 19, 20, 21, 22, 23, 24])\n",
        "10.1371/journal.pbio.0000005    set([10, 25, 26, 27, 28, 29, 30, 31])\n",
        "Name: subj_leaf, dtype: object\n",
        "id\n",
        "10.1371/journal.pbio.0000001    True\n",
        "10.1371/journal.pbio.0000002    True\n",
        "10.1371/journal.pbio.0000003    True\n",
        "10.1371/journal.pbio.0000004    True\n",
        "10.1371/journal.pbio.0000005    True\n",
        "Name: subj_leaf, dtype: bool\n"
       ]
      }
     ],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Test again!\n",
      "\n",
      "print test_df['subject'].apply(len)\n",
      "\n",
      "d_to_code, d_from_code = make_dicts(test_df['subject'])\n",
      "test_enc2 = encode_terms(test_df['subject'], d_to_code)\n",
      "print test_enc2\n",
      "print test_enc2.apply(len) == test_df['subject'].apply(len)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "id\n",
        "10.1371/journal.pbio.0000001    13\n",
        "10.1371/journal.pbio.0000002    14\n",
        "10.1371/journal.pbio.0000003    18\n",
        "10.1371/journal.pbio.0000004    16\n",
        "10.1371/journal.pbio.0000005    12\n",
        "Name: subject, dtype: int64\n",
        "id\n",
        "10.1371/journal.pbio.0000001     set([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13])\n",
        "10.1371/journal.pbio.0000002    set([5, 9, 10, 12, 14, 15, 16, 17, 18, 19, 20,...\n",
        "10.1371/journal.pbio.0000003    set([32, 33, 5, 12, 15, 16, 17, 18, 20, 21, 24...\n",
        "10.1371/journal.pbio.0000004    set([34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 4...\n",
        "10.1371/journal.pbio.0000005    set([54, 47, 48, 49, 50, 51, 52, 53, 22, 55, 5...\n",
        "Name: subject, dtype: object\n",
        "id\n",
        "10.1371/journal.pbio.0000001    True\n",
        "10.1371/journal.pbio.0000002    True\n",
        "10.1371/journal.pbio.0000003    True\n",
        "10.1371/journal.pbio.0000004    True\n",
        "10.1371/journal.pbio.0000005    True\n",
        "Name: subject, dtype: bool\n"
       ]
      }
     ],
     "prompt_number": 9
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Test output dict as JSON:\n",
      "json.dumps(d_from_code)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 10,
       "text": [
        "'{\"1\": \"/Biology and life sciences/Molecular biology/Molecular biology techniques/Molecular biology assays and analysis techniques/Library screening/Recombination-based assay\", \"2\": \"/Research and analysis methods/Molecular biology techniques/Sequencing techniques/Sequence analysis/DNA sequence analysis\", \"3\": \"/Physical sciences/Chemistry/Chemical reactions/Recombination reactions\", \"4\": \"/Biology and life sciences/Molecular biology/Molecular biology techniques/Sequencing techniques/Nucleotide sequencing\", \"5\": \"/Biology and life sciences/Biochemistry/DNA/DNA sequences\", \"6\": \"/Biology and life sciences/Biochemistry/Proteins/Recombinant proteins\", \"7\": \"/Biology and life sciences/Molecular biology/Molecular biology techniques/Sequencing techniques/Sequence analysis/DNA sequence analysis\", \"8\": \"/Biology and life sciences/Biochemistry/Nucleotides\", \"9\": \"/Research and analysis methods/Molecular biology techniques/Sequencing techniques/Sequence analysis/Sequence motif analysis\", \"10\": \"/Biology and life sciences/Molecular biology/Molecular biology techniques/Sequencing techniques/Sequence analysis/Sequence motif analysis\", \"11\": \"/Research and analysis methods/Molecular biology techniques/Molecular biology assays and analysis techniques/Library screening/Recombination-based assay\", \"12\": \"/Biology and life sciences/Genetics/DNA/DNA sequences\", \"13\": \"/Research and analysis methods/Molecular biology techniques/Sequencing techniques/Nucleotide sequencing\", \"14\": \"/Biology and life sciences/Molecular biology/Molecular biology techniques/Artificial gene amplification and extension/Polymerase chain reaction\", \"15\": \"/Biology and life sciences/Organisms/Viruses/Viral pathogens/Coronaviruses\", \"16\": \"/Medicine and health sciences/Pathology and laboratory medicine/Pathogens/Microbial pathogens/Viral pathogens/Coronaviruses\", \"17\": \"/Biology and life sciences/Microbiology/Medical microbiology/Microbial pathogens/Viral pathogens/Coronaviruses\", \"18\": \"/Medicine and health sciences/Infectious diseases/Viral diseases/SARS\", \"19\": \"/Research and analysis methods/Molecular biology techniques/Artificial gene amplification and extension/Polymerase chain reaction\", \"20\": \"/Biology and life sciences/Organisms/Viruses/RNA viruses/Coronaviruses\", \"21\": \"/Research and analysis methods/Bioassays and physiological analysis/Microarrays\", \"22\": \"/Physical sciences/Materials science/Materials by structure/Polymers/Oligonucleotides\", \"23\": \"/Biology and life sciences/Biochemistry/Nucleic acids\", \"24\": \"/Biology and life sciences/Organisms/Viruses/Viral pathogens/Flaviviruses/West Nile virus\", \"25\": \"/Research and analysis methods/Molecular biology techniques/Sequencing techniques/Sequence analysis\", \"26\": \"/Biology and life sciences/Microbiology/Medical microbiology/Microbial pathogens/Viral pathogens/Flaviviruses/West Nile virus\", \"27\": \"/Medicine and health sciences/Pathology and laboratory medicine/Pathogens/Microbial pathogens/Viral pathogens/Flaviviruses/West Nile virus\", \"28\": \"/Biology and life sciences/Microbiology/Microbial genomics/Viral genomics\", \"29\": \"/Biology and life sciences/Molecular biology/Molecular biology techniques/Sequencing techniques/Sequence analysis\", \"30\": \"/Biology and life sciences/Genetics/Genomics/Microbial genomics/Viral genomics\", \"31\": \"/Biology and life sciences/Microbiology/Virology/Viral genomics\", \"32\": \"/Biology and life sciences/Organisms/Viruses/RNA viruses/Flaviviruses/West Nile virus\", \"33\": \"/Biology and life sciences/Organisms/Viruses/DNA viruses\", \"34\": \"/Medicine and health sciences/Immunology/Immune system proteins\", \"35\": \"/Biology and life sciences/Biochemistry/Proteins/Protein interactions\", \"36\": \"/Medicine and health sciences/Physiology/Immune physiology/Antibodies\", \"37\": \"/Biology and life sciences/Immunology/Immune system proteins/Antibodies\", \"38\": \"/Biology and life sciences/Physiology/Immune physiology/Antibodies\", \"39\": \"/Biology and life sciences/Biochemistry/Proteins/Immune system proteins/Antibodies\", \"40\": \"/Biology and life sciences/Immunology/Immune system proteins\", \"41\": \"/Medicine and health sciences/Immunology/Immune system proteins/Antibodies\", \"42\": \"/Biology and life sciences/Biochemistry/Proteins/Immune system proteins\", \"43\": \"/Medicine and health sciences\", \"44\": \"/Biology and life sciences/Biochemistry/DNA/DNA recombination\", \"45\": \"/Biology and life sciences/Genetics/Gene expression/Gene regulation\", \"46\": \"/Biology and life sciences/Genetics/DNA/DNA recombination\", \"47\": \"/Biology and life sciences/Biochemistry/Proteins/Enzymes/Proteases\", \"48\": \"/Biology and life sciences/Parasitology/Parasite groups/Apicomplexa/Merozoites\", \"49\": \"/Biology and life sciences/Biochemistry/Enzymology/Enzymes/Proteases\", \"50\": \"/Biology and life sciences/Computational biology/Genome analysis/Transcriptome analysis\", \"51\": \"/Biology and life sciences/Genetics/Genomics/Genome analysis/Gene prediction\", \"52\": \"/Biology and life sciences/Computational biology/Genome analysis/Gene prediction\", \"53\": \"/Biology and life sciences/Genetics/Gene expression\", \"54\": \"/Biology and life sciences/Genetics/Genomics/Genome analysis/Transcriptome analysis\", \"55\": \"/Biology and life sciences/Parasitology/Parasite groups/Apicomplexa/Plasmodium\", \"56\": \"/Biology and life sciences/Cell biology/Plant cell biology/Plastids\", \"57\": \"/Biology and life sciences/Plant science/Plant cell biology/Plastids\"}'"
       ]
      }
     ],
     "prompt_number": 10
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Ready to generate dicts & translate the entire dataset\n",
      "\n",
      "This cell takes a long time to run."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# subject\n",
      "subject_enc, subject_dec = make_dicts(df['subject'])\n",
      "\n",
      "with open('../data/dict_subject_enc.json', 'wb') as f:\n",
      "     json.dump(subject_enc, f)\n",
      "\n",
      "with open('../data/dict_subject_dec.json', 'wb') as f:\n",
      "     json.dump(subject_dec, f)\n",
      "\n",
      "# subj_top\n",
      "subj_top_enc, subj_top_dec = make_dicts(df['subj_top'])\n",
      "\n",
      "with open('../data/dict_subj_top.json', 'wb') as f:\n",
      "     json.dump(subj_top_dec, f)\n",
      "\n",
      "# subj_leaf\n",
      "subj_leaf_enc, subj_leaf_dec = make_dicts(df['subj_leaf'])\n",
      "\n",
      "with open('../data/dict_subj_leaf.json', 'wb') as f:\n",
      "     json.dump(subj_leaf_dec, f)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 11
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "The next cell also takes a while."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df['subject'] = encode_terms(df['subject'], subject_enc)\n",
      "df['subj_top'] = encode_terms(df['subj_top'], subj_top_enc)\n",
      "df['subj_leaf'] = encode_terms(df['subj_leaf'], subj_leaf_enc)\n",
      "\n",
      "df.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>publication_date</th>\n",
        "      <th>subject</th>\n",
        "      <th>subj_top</th>\n",
        "      <th>subj_leaf</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>id</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>10.1371/journal.pone.0100281</th>\n",
        "      <td> 2014</td>\n",
        "      <td> set([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13...</td>\n",
        "      <td>       set([1, 2, 3])</td>\n",
        "      <td>         set([1, 2, 3, 4, 5, 6, 7, 8])</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>10.1371/journal.pone.0070935</th>\n",
        "      <td> 2013</td>\n",
        "      <td> set([32, 17, 18, 19, 20, 21, 22, 23, 24, 25, 2...</td>\n",
        "      <td> set([1, 2, 3, 4, 5])</td>\n",
        "      <td>  set([9, 10, 11, 12, 13, 14, 15, 16])</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>10.1371/journal.pone.0066671</th>\n",
        "      <td> 2013</td>\n",
        "      <td> set([33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43])</td>\n",
        "      <td>    set([8, 1, 6, 7])</td>\n",
        "      <td> set([17, 18, 19, 20, 21, 22, 23, 24])</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>10.1371/journal.pntd.0002444</th>\n",
        "      <td> 2013</td>\n",
        "      <td>     set([44, 45, 46, 47, 48, 49, 50, 51, 52, 53])</td>\n",
        "      <td>       set([2, 5, 7])</td>\n",
        "      <td> set([32, 25, 26, 27, 28, 29, 30, 31])</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>10.1371/journal.pone.0003605</th>\n",
        "      <td> 2008</td>\n",
        "      <td>         set([54, 55, 56, 57, 58, 59, 60, 61, 62])</td>\n",
        "      <td>       set([1, 2, 3])</td>\n",
        "      <td> set([33, 34, 35, 36, 37, 38, 39, 40])</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>5 rows \u00d7 4 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 12,
       "text": [
        "                             publication_date  \\\n",
        "id                                              \n",
        "10.1371/journal.pone.0100281             2014   \n",
        "10.1371/journal.pone.0070935             2013   \n",
        "10.1371/journal.pone.0066671             2013   \n",
        "10.1371/journal.pntd.0002444             2013   \n",
        "10.1371/journal.pone.0003605             2008   \n",
        "\n",
        "                                                                        subject  \\\n",
        "id                                                                                \n",
        "10.1371/journal.pone.0100281  set([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13...   \n",
        "10.1371/journal.pone.0070935  set([32, 17, 18, 19, 20, 21, 22, 23, 24, 25, 2...   \n",
        "10.1371/journal.pone.0066671  set([33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43])   \n",
        "10.1371/journal.pntd.0002444      set([44, 45, 46, 47, 48, 49, 50, 51, 52, 53])   \n",
        "10.1371/journal.pone.0003605          set([54, 55, 56, 57, 58, 59, 60, 61, 62])   \n",
        "\n",
        "                                          subj_top  \\\n",
        "id                                                   \n",
        "10.1371/journal.pone.0100281        set([1, 2, 3])   \n",
        "10.1371/journal.pone.0070935  set([1, 2, 3, 4, 5])   \n",
        "10.1371/journal.pone.0066671     set([8, 1, 6, 7])   \n",
        "10.1371/journal.pntd.0002444        set([2, 5, 7])   \n",
        "10.1371/journal.pone.0003605        set([1, 2, 3])   \n",
        "\n",
        "                                                          subj_leaf  \n",
        "id                                                                   \n",
        "10.1371/journal.pone.0100281          set([1, 2, 3, 4, 5, 6, 7, 8])  \n",
        "10.1371/journal.pone.0070935   set([9, 10, 11, 12, 13, 14, 15, 16])  \n",
        "10.1371/journal.pone.0066671  set([17, 18, 19, 20, 21, 22, 23, 24])  \n",
        "10.1371/journal.pntd.0002444  set([32, 25, 26, 27, 28, 29, 30, 31])  \n",
        "10.1371/journal.pone.0003605  set([33, 34, 35, 36, 37, 38, 39, 40])  \n",
        "\n",
        "[5 rows x 4 columns]"
       ]
      }
     ],
     "prompt_number": 12
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Here is where we check to see if I did it right..."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df.head().to_json(orient='index', force_ascii=False)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 13,
       "text": [
        "'{\"10.1371\\\\/journal.pone.0100281\":{\"publication_date\":\"2014\",\"subject\":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16],\"subj_top\":[1,2,3],\"subj_leaf\":[1,2,3,4,5,6,7,8]},\"10.1371\\\\/journal.pone.0070935\":{\"publication_date\":\"2013\",\"subject\":[32,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31],\"subj_top\":[1,2,3,4,5],\"subj_leaf\":[9,10,11,12,13,14,15,16]},\"10.1371\\\\/journal.pone.0066671\":{\"publication_date\":\"2013\",\"subject\":[33,34,35,36,37,38,39,40,41,42,43],\"subj_top\":[8,1,6,7],\"subj_leaf\":[17,18,19,20,21,22,23,24]},\"10.1371\\\\/journal.pntd.0002444\":{\"publication_date\":\"2013\",\"subject\":[44,45,46,47,48,49,50,51,52,53],\"subj_top\":[2,5,7],\"subj_leaf\":[32,25,26,27,28,29,30,31]},\"10.1371\\\\/journal.pone.0003605\":{\"publication_date\":\"2008\",\"subject\":[54,55,56,57,58,59,60,61,62],\"subj_top\":[1,2,3],\"subj_leaf\":[33,34,35,36,37,38,39,40]}}'"
       ]
      }
     ],
     "prompt_number": 13
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## If all is OK, export.."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df.to_json(path_or_buf='../data/articles_coded.json', orient='index', force_ascii=False)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 14
    }
   ],
   "metadata": {}
  }
 ]
}